src/lib/pubkey/hss_lms/lms.cpp

```diff
@@ -0,0 +1,351 @@
+   }
+   if(type >= LMS_Algorithm_Type::SHAKE_M32_H5 && type <= LMS_Algorithm_Type::SHAKE_M32_H25) {
+      uint8_t h = 5 * (type_value - checked_cast_to<uint8_t>(LMS_Algorithm_Type::SHAKE_M32_H5) + 1);
+      return LMS_Params(type, "SHAKE-256(256)", h);
+   }
+   if(type >= LMS_Algorithm_Type::SHAKE_M24_H5 && type <= LMS_Algorithm_Type::SHAKE_M24_H25) {
+      uint8_t h = 5 * (type_value - checked_cast_to<uint8_t>(LMS_Algorithm_Type::SHAKE_M24_H5) + 1);
+      return LMS_Params(type, "SHAKE-256(192)", h);
+   }
+
+   throw Decoding_Error("Unsupported LMS algorithm type");
+}
+
+LMS_Params LMS_Params::create_or_throw(std::string_view hash_name, uint8_t h) {
+   BOTAN_ARG_CHECK(h == 5 || h == 10 || h == 15 || h == 20 || h == 25, "Invalid h value");
+   auto type_offset = h / 5 - 1;
+   LMS_Algorithm_Type base_type;
+
+   if(hash_name == "SHA-256") {
+      base_type = LMS_Algorithm_Type::SHA256_M32_H5;
+   } else if(hash_name == "Truncated(SHA-256,192)") {
+      base_type = LMS_Algorithm_Type::SHA256_M24_H5;
+   } else if(hash_name == "SHAKE-256(256)") {
+      base_type = LMS_Algorithm_Type::SHAKE_M32_H5;
+   } else if(hash_name == "SHAKE-256(192)") {
+      base_type = LMS_Algorithm_Type::SHAKE_M24_H5;
+   } else {
+      throw Decoding_Error("Unsupported hash function");
+   }
+   auto type = checked_cast_to<LMS_Algorithm_Type>(checked_cast_to<uint8_t>(base_type) + type_offset);
+   return LMS_Params(type, hash_name, h);
+}
+
+LMS_Params::LMS_Params(LMS_Algorithm_Type algorithm_type, std::string_view hash_name, uint8_t h) :
+      m_algorithm_type(algorithm_type), m_h(h), m_hash_name(hash_name) {
+   const auto hash = HashFunction::create_or_throw(m_hash_name);
+   m_m = hash->output_length();
+}
+
+LMS_PublicKey LMS_PrivateKey::sign_and_get_pk(StrongSpan<LMS_Signature_Bytes> out_sig,
+                                              LMS_Tree_Node_Idx q,
+                                              const LMS_Message& msg) const {
+   // Pre-alloc space for the signature
+   BOTAN_ARG_CHECK(out_sig.size() == LMS_Signature::size(lms_params(), lmots_params()), "Invalid output buffer size");
+
+   BufferStuffer sig_stuffer(out_sig);
+   sig_stuffer.append_be(q);
+   const LMOTS_Private_Key lmots_sk(lmots_params(), identifier(), q, seed());
+   lmots_sk.sign(sig_stuffer.next<LMOTS_Signature_Bytes>(LMOTS_Signature::size(lmots_params())), msg);
+   sig_stuffer.append_be(lms_params().algorithm_type());
+   const auto auth_path_buffer = sig_stuffer.next<LMS_AuthenticationPath>(lms_params().m() * lms_params().h());
+
+   BOTAN_ASSERT_NOMSG(sig_stuffer.full());
+
+   TreeAddress lms_tree_address(lms_params().h());
+   LMS_Tree_Node pk_buffer(lms_params().m());
+   lms_treehash(StrongSpan<LMS_Tree_Node>(pk_buffer.get()), auth_path_buffer, q, *this);
+
+   return LMS_PublicKey(lms_params(), lmots_params(), identifier(), std::move(pk_buffer));
+}
+
+LMS_PublicKey LMS_PublicKey::from_bytes_of_throw(BufferSlicer& slicer) {
```

Comment:
```suggestion
LMS_PublicKey LMS_PublicKey::from_bytes_or_throw(BufferSlicer& slicer) {
```

Please don't forget the header.